### PR TITLE
fix creation and deletion timestamp for work metadata.

### PR DESCRIPTION
**Pull request by morvencao**
fix: https://github.com/openshift-online/maestro/issues/158

by setting the CreationTimestamp and DeletionTimestamp when publishing the manifest work and saving them to the database.

when deletion is triggered, the resource-bundles API returns:
```
{
  "items": [
    {
      "consumer_name": "fed01f98-4d6e-4f21-9d8a-fb70831dc987",
      "created_at": "2024-07-11T08:12:19.362465Z",
      "delete_option": {},
      "deleted_at": "2024-07-11T08:12:44.400699Z",
      "href": "/api/maestro/v1/resource-bundles/39fb0720-6487-582d-ba1f-1df6be62d427",
      "id": "39fb0720-6487-582d-ba1f-1df6be62d427",
      "kind": "ResourceBundle",
      "manifest_configs": [],
      "manifests": [
        {
          "apiVersion": "v1",
          "data": {
            "test": "wxclm"
          },
          "kind": "ConfigMap",
          "metadata": {
            "name": "work-k6lbc",
            "namespace": "default"
          }
        }
      ],
      "metadata": {
        "annotations": {
          "work.annotations": "example"
        },
        "creationTimestamp": "2024-07-11T08:12:19Z",
        "deletionTimestamp": "2024-07-11T08:12:44Z",
        "labels": {
          "work.label": "example"
        },
        "name": "work-k6lbc",
        "namespace": "fed01f98-4d6e-4f21-9d8a-fb70831dc987",
        "resourceVersion": "1",
        "uid": "39fb0720-6487-582d-ba1f-1df6be62d427"
      },
```

### CODE REVIEW

File: cmd/maestro/server/grpc_server.go [hunk 1]
```diff
@@ -150,8 +176,22 @@ func (svr *GRPCServer) Publish(ctx context.Context, pubReq *pbv1.PublishRequest)
...
 	case common.DeleteRequestAction:
-		err := svr.resourceService.MarkAsDeleting(ctx, res.ID)
+		res, err = decode(eventType.CloudEventsDataType, evt)
 		if err != nil {
+			return nil, fmt.Errorf("failed to decode cloudevent: %v", err)
+		}
+		var deletionTimestamp time.Time
+		evtExtensions := evt.Context.GetExtensions()
+		if _, ok := evtExtensions[codec.ExtensionWorkMeta]; ok {
+			if deletionTimestampValue, exists := evtExtensions[types.ExtensionDeletionTimestamp]; exists {
```
Comment (skeeey) [inline cmd/maestro/server/grpc_server.go:186]: should we ignore the client deletion timestamp? instead of setting the deletion timestamp on the server

File: cmd/maestro/server/grpc_server.go [hunk 2]
```diff
@@ -150,8 +176,19 @@ func (svr *GRPCServer) Publish(ctx context.Context, pubReq *pbv1.PublishRequest)
...
+		res, err = decode(eventType.CloudEventsDataType, evt)
 		if err != nil {
+			return nil, fmt.Errorf("failed to decode cloudevent: %v", err)
+		}
+		var deletionTimestamp time.Time
+		evtExtensions := evt.Context.GetExtensions()
+		if _, ok := evtExtensions[codec.ExtensionWorkMeta]; ok {
+			if _, exists := evtExtensions[types.ExtensionDeletionTimestamp]; exists {
+				deletionTimestamp = time.Now()
+			}
```
Comment (machi1990) [inline cmd/maestro/server/grpc_server.go:188]: @skeeey is `evtExtensions[types.ExtensionDeletionTimestamp]` set by the source client always when deleting a manifest work?

File: cmd/maestro/server/grpc_server.go (hunk 1 above)
Comment (morvencao) [reply cmd/maestro/server/grpc_server.go]: done, use server generated deletion timestamp.

File: cmd/maestro/server/grpc_server.go (hunk 2 above)
Comment (morvencao) [reply cmd/maestro/server/grpc_server.go:188]: yes, it is.

Comment (machi1990) [reply cmd/maestro/server/grpc_server.go:188]: Thanks for confirming.

File: cmd/maestro/server/grpc_server.go [hunk 3]
```diff
@@ -150,8 +176,19 @@ func (svr *GRPCServer) Publish(ctx context.Context, pubReq *pbv1.PublishRequest)
...
-		err := svr.resourceService.MarkAsDeleting(ctx, res.ID)
+		res, err = decode(eventType.CloudEventsDataType, evt)
 		if err != nil {
+			return nil, fmt.Errorf("failed to decode cloudevent: %v", err)
+		}
+		var deletionTimestamp time.Time
+		evtExtensions := evt.Context.GetExtensions()
+		if _, ok := evtExtensions[codec.ExtensionWorkMeta]; ok {
+			if _, exists := evtExtensions[types.ExtensionDeletionTimestamp]; exists {
+				deletionTimestamp = time.Now()
```
Comment (machi1990) [inline cmd/maestro/server/grpc_server.go:187]: we are not setting 
```go
 res.Meta.DeletedAt = gorm.DeletedAt{
	Time:  deletionTimestamp,
        Valid: true,
}
```

just like what we do for create to ensure that the two  field values (the timestamp in metadata and in resource bundle) are the same.

File: pkg/services/resource.go [hunk 4]
```diff
@@ -214,16 +215,41 @@ func (s *sqlResourceService) UpdateStatus(ctx context.Context, resource *api.Res
...
+		}
+		// set deletionTimestamp in metadata
+		workMetaValue, ok := found.Payload["metadata"]
+		if ok {
+			workMeta, ok := workMetaValue.(map[string]interface{})
+			if ok {
+				workMeta["deletionTimestamp"] = deletionTimestamp.Format(time.RFC3339)
+				found.Payload["metadata"] = workMeta
+			}
+		}
```
Comment (machi1990) [inline pkg/services/resource.go:245]: The comment in https://github.com/openshift-online/maestro/pull/159/files#r1673549425 applies here I believe.

File: cmd/maestro/server/grpc_server.go (hunk 3 above)
Comment (machi1990) [reply cmd/maestro/server/grpc_server.go:187]: Please see https://github.com/openshift-online/maestro/pull/159/files#r1673551640

Comment (morvencao) [reply cmd/maestro/server/grpc_server.go:187]: I encountered to apply the `deleteAt` for resource using `Update`, let me try again when we have `Valid: true`.

Comment (machi1990) [reply cmd/maestro/server/grpc_server.go:187]: Thank you for looking. Please note, this should be done in the line comment at https://github.com/openshift-online/maestro/pull/159#discussion_r1673551640

File: pkg/services/resource.go [hunk 5]
```diff
@@ -214,16 +215,41 @@ func (s *sqlResourceService) UpdateStatus(ctx context.Context, resource *api.Res
...
+		_, err = s.resourceDao.Update(ctx, found)
+		if err != nil {
+			return handleUpdateError("Resource", err)
+		}
+		if err := s.resourceDao.Delete(ctx, id, false); err != nil {
+			return handleDeleteError("Resource", errors.GeneralError("Unable to delete resource: %s", err))
+		}
```
[the remark this answers is not in the document]
Comment (morvencao) [reply pkg/services/resource.go]: removed the `Delete` operation as the `deleteAt` is updated using `Update` operation.

File: cmd/maestro/server/grpc_server.go (hunk 3 above)
Comment (morvencao) [reply cmd/maestro/server/grpc_server.go:188]: Looks like the `valid: true` work.
I updated this so that work meta and resource `deleteAt` are using the same `deletionTimestamp`.

**Review by machi1990 — COMMENTED** [review]
lgtm

Might be useful to add an integration test for this

File: pkg/services/resource.go [hunk 6]
```diff
@@ -214,16 +216,40 @@ func (s *sqlResourceService) UpdateStatus(ctx context.Context, resource *api.Res
...
 // 3. Maestro handles delete event and sends CloudEvent to work-agent
 // 4. Work-agent deletes resource, sends CloudEvent back to Maestro
 // 5. Maestro hard deletes resource from DB
-func (s *sqlResourceService) MarkAsDeleting(ctx context.Context, id string) *errors.ServiceError {
-	if err := s.resourceDao.Delete(ctx, id, false); err != nil {
-		return handleDeleteError("Resource", errors.GeneralError("Unable to delete resource: %s", err))
+func (s *sqlResourceService) MarkAsDeleting(ctx context.Context, id string, deletionTimestamp time.Time) *errors.ServiceError {
+	if deletionTimestamp.IsZero() {
+		// if deletionTimestamp is not provided, delete the resource immediately
+		if err := s.resourceDao.Delete(ctx, id, false); err != nil {
```
[the remark this answers is not in the document]
Comment (morvencao) [reply pkg/services/resource.go:229]: do you mean case `deletionTimestamp.IsZero()`? this case, we don't have work metadata.

File: cmd/maestro/server/grpc_server.go [hunk 7]
```diff
@@ -150,8 +176,20 @@ func (svr *GRPCServer) Publish(ctx context.Context, pubReq *pbv1.PublishRequest)
...
 	case common.DeleteRequestAction:
-		err := svr.resourceService.MarkAsDeleting(ctx, res.ID)
+		res, err = decode(eventType.CloudEventsDataType, evt)
 		if err != nil {
+			return nil, fmt.Errorf("failed to decode cloudevent: %v", err)
+		}
+		var deletionTimestamp time.Time
+		evtExtensions := evt.Context.GetExtensions()
+		if _, ok := evtExtensions[codec.ExtensionWorkMeta]; ok {
+			if _, exists := evtExtensions[types.ExtensionDeletionTimestamp]; exists {
```
[the remark this answers is not in the document]
Comment (morvencao) [reply cmd/maestro/server/grpc_server.go:186]: from source side, the work client set this in codec: https://github.com/open-cluster-management-io/sdk-go/blob/main/pkg/cloudevents/work/source/codec/manifestbundle.go#L58

File: pkg/services/resource.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (morvencao) [reply pkg/services/resource.go:236]: if deletionTimestamp is zero, this case is for deleting resource without work metadata, just following original flow use database soft delete.

File: cmd/maestro/server/grpc_server.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (clyang82) [reply cmd/maestro/server/grpc_server.go:186]: the comment is clear for me to understand. Thanks.

**Review by clyang82 — APPROVED** [review]
LGTM
